### PR TITLE
REACTOR-289 - Reactor support for CloudZero Normalized Event Format

### PR DESCRIPTION
**Pull request by tiyberius**

### CODE REVIEW

File: reactor/aws/cloudtrail.py [hunk 1]
```diff
@@ -372,3 +399,281 @@ def traverse_map(obj, map):
...
+    >>> get_arn_from_user_identity(None)
+    """
+    try:
+        account_id = user_identity_field.get('accountId')
+        resource_id = get_resource_id_from_principal_id(user_identity_field.get('principalId'))
+        resource_type, resource_service = determine_resource_type_and_service(resource_id)
+        service = resource_service or service
+        arn = user_identity_field.get('arn')
+        if arn is None:
+            if service and region and account_id and resource_type and resource_id:
```
Comment (tiyberius) [inline reactor/aws/cloudtrail.py]: MAJOR Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=python%3AS1066)

**Comment by tiyberius** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: reactor/data/aws/datamaps/event_default_mappings.py [hunk 2]
```diff
@@ -19,11 +19,7 @@
       "last_used":"2017-05-03T15:10:21.951721+0000",
       "count":0,
       "map":{
-         "userIdentity":{
-            "type":"event_identity_type",
-            "accountId":"event_actor_account_id",
-            "principalId":"event_actor_principal_id"
-         }
+         "userIdentity":{}
```
Comment (silvexis) [inline reactor/data/aws/datamaps/event_default_mappings.py]: I don't understand these maps, you are essentially throwing away these fields when you map them to {} aren't you? That seems very wrong.

File: test/integration/aws/test_integration_cloudtrail.py [hunk 3]
```diff
@@ -23,20 +23,10 @@ def test_mapping_single_event(event_data, reactor_event_mapping_table,
...
-                }
-            },
-            'accountId': 'event_actor_account_id',
-            'principalId': 'event_actor_principal_id',
-            'invokedBy': 'event_actor_invoker',
-            'type': 'event_identity_type',
-            'arn': 'event_actor_arn'
+        "userIdentity": {
+            "sessionContext": {
+                "sessionIssuer": {}
```
Comment (silvexis) [inline test/integration/aws/test_integration_cloudtrail.py]: this um kinda tests nothing?

File: reactor/aws/cloudtrail.py [hunk 4]
```diff
@@ -258,45 +259,71 @@ def normalize(self, environments, event_owner_id=None, event_ttl=None):
...
+            all_arns.append(target_resource_arn)
+        if event_source_arn:
+            all_arns.append(event_source_arn)
         self.normalized_event = {
-            'event_id': self.raw_event['eventID'],
+            'event_id': raw_event['eventID'],
+            'event_source_type': 'CLOUDTRAIL',
+            'event_name': raw_event['eventName'],
+            'event_type': raw_event['eventType'],
             'environment_id': environment.get('environment_id'),
```
Comment (silvexis) [inline reactor/aws/cloudtrail.py]: i propose you make this break if environment_id isn't there, we want to ensure an env_id is aways in the record.

File: reactor/aws/cloudtrail.py [hunk 5]
```diff
@@ -258,45 +259,71 @@ def normalize(self, environments, event_owner_id=None, event_ttl=None):
...
-            'aws_account_id': self.raw_event['recipientAccountId'],
-            'source_env_id': self.event_owner_id,
-            'event_datetime': self.raw_event['eventTime'],
+            'environment_type': 'AWS',
+            'cloud_account_id': raw_event['recipientAccountId'],
+            # todo handle all types of events that do not have an arn in userIdentity
+            'event_actor_resource_id': event_source_arn,
+            # todo determine if the source IP address is from the internet or from within an AWS account
+            # 'event_source_resource_id': construct_czrn(
```
Comment (silvexis) [inline reactor/aws/cloudtrail.py]: commented out?

File: reactor/aws/cloudtrail.py [hunk 6]
```diff
@@ -372,3 +399,281 @@ def traverse_map(obj, map):
...
+    >>> construct_czrn('aws', 'ec2', 'us-east-1', '123456789012', 'ipv4_address', '10.0.1.45', '/24')
+    'czrn:aws:ec2:us-east-1:123456789012:ipv4_address:10.0.1.45/24'
+    >>> construct_czrn('aws', 'ec2', 'us-east-1', '123456789012', 'ipv4_address', '10.0.1.45')
+    'czrn:aws:ec2:us-east-1:123456789012:ipv4_address:10.0.1.45'
+    """
+    resource_plus_path = f'{resource}{path}' if path else f'{resource}'
+    return ':'.join(['czrn', provider, service, region, account_id, resource_type, resource_plus_path])
+
+
+def construct_arn(service, region, account_id, resource_type, resource):
```
Comment (silvexis) [inline reactor/aws/cloudtrail.py]: Return False vs "Failed to construct ARN"?

File: reactor/aws/cloudtrail.py (hunk 1 above)
Comment (tiyberius) [reply reactor/aws/cloudtrail.py]: okayyyyy fffiiiinne

File: reactor/aws/cloudtrail.py (hunk 4 above)
Comment (tiyberius) [reply reactor/aws/cloudtrail.py]: done

File: reactor/aws/cloudtrail.py (hunk 5 above)
Comment (tiyberius) [reply reactor/aws/cloudtrail.py]: deeeeeleted

File: reactor/aws/cloudtrail.py (hunk 6 above)
Comment (tiyberius) [reply reactor/aws/cloudtrail.py]: done

Comment (tiyberius) [reply reactor/aws/cloudtrail.py]: this is done

**Pull request merged**